File: backend/app/services/form_service.py
"""Business logic for Form operations."""

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.crud import form as form_crud
from app.schemas.form import FormCreate, FormUpdate


def list_forms(db: Session, skip: int = 0, limit: int = 100):
    """List all forms with response counts."""
    forms = form_crud.get_forms(db, skip=skip, limit=limit)
    return forms


def get_form_detail(db: Session, form_id: int):
    """Get form with questions and response count. Raises 404 if not found."""
    form = form_crud.get_form(db, form_id)
    if not form:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Form with id {form_id} not found",
        )
    return form


def create_form(db: Session, data: FormCreate):
    """Create a new form."""
    return form_crud.create_form(db, data)


def update_form(db: Session, form_id: int, data: FormUpdate):
    """Update a form. Raises 404 if not found."""
    form = form_crud.update_form(db, form_id, data)
    if not form:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Form with id {form_id} not found",
        )
    return form


def delete_form(db: Session, form_id: int):
    """Delete a form. Raises 404 if not found."""
    success = form_crud.delete_form(db, form_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Form with id {form_id} not found",
        )
    return True


def duplicate_form(db: Session, form_id: int):
    """Duplicate a form and its questions. Raises 404 if not found."""
    form = form_crud.duplicate_form(db, form_id)
    if not form:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Form with id {form_id} not found",
        )
    return form


def publish_form(db: Session, form_id: int):
    """Publish a form. Must have at least 1 question."""
    form = form_crud.get_form(db, form_id)
    if not form:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Form with id {form_id} not found",
        )
    if len(form.questions) == 0:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot publish a form with no questions",
        )
    form_update = FormUpdate(status="published")
    return form_crud.update_form(db, form_id, form_update)


def unpublish_form(db: Session, form_id: int):
    """Unpublish a form (set back to draft)."""
    form = form_crud.get_form(db, form_id)
    if not form:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Form with id {form_id} not found",
        )
    form_update = FormUpdate(status="draft")
    return form_crud.update_form(db, form_id, form_update)

File: frontend/src/services/formService.ts
/**
 * Form Service — all API calls related to forms.
 * 
 * This is the "service layer" for the frontend.
 * Components never call axios directly — they call these functions instead.
 * This keeps API logic in one place and makes it easy to change endpoints.
 */

import api from "@/lib/api";
import type { Form, FormDetail, FormCreate, FormUpdate } from "@/types/form";

export const formService = {
  /** Get all forms */
  async getForms(): Promise<Form[]> {
    const { data } = await api.get<Form[]>("/forms/");
    return data;
  },

  /** Get a single form with questions and response count */
  async getForm(id: number): Promise<FormDetail> {
    const { data } = await api.get<FormDetail>(`/forms/${id}`);
    return data;
  },

  /** Create a new form */
  async createForm(form: FormCreate): Promise<Form> {
    const { data } = await api.post<Form>("/forms/", form);
    return data;
  },

  /** Update a form (title, description, status) */
  async updateForm(id: number, form: FormUpdate): Promise<Form> {
    const { data } = await api.put<Form>(`/forms/${id}`, form);
    return data;
  },

  /** Delete a form */
  async deleteForm(id: number): Promise<void> {
    await api.delete(`/forms/${id}`);
  },

  /** Duplicate a form */
  async duplicateForm(id: number): Promise<Form> {
    const { data } = await api.post<Form>(`/forms/${id}/duplicate`);
    return data;
  },

  /** Publish a form */
  async publishForm(id: number): Promise<Form> {
    const { data } = await api.post<Form>(`/forms/${id}/publish`);
    return data;
  },

  /** Unpublish a form */
  async unpublishForm(id: number): Promise<Form> {
    const { data } = await api.post<Form>(`/forms/${id}/unpublish`);
    return data;
  },
};
